Добавил функционал жалоб

StDemidov opened the pull request:
Реализовал функционал жалоб пока только на пользователей
- Реализовал возможность прикрепления скриншотов
- Поменял dataclass Report + enum Categories (добавил поле скриншота и поменял категории)
- Пока не создавал в конце запрос к апи на создание жалобы

StDemidov committed:
Add create_report to APIService, add Categories enum and Report dataclass

StDemidov committed:
Add complain

TomatoInOil commented on src/bot/conversations/complain/callback_funcs.py:
```diff
@@ -0,0 +1,150 @@
+        await update.effective_chat.send_message(
+            text=message_text,
+            reply_markup=ReplyKeyboardRemove(),
+        )
```
![image](https://github.com/Studio-Yandex-Practicum/coliving_bot/assets/93451214/9ef185d6-f2d7-4451-8bf6-d4f0399d0986)

Т.к. кнопка "Нет, у меня нет скриншота" - это inline кнопка, то ReplyKeyboardRemove() не работает в этом случае. В этом случае можно использовать 
```Python
# Заменит текст сообщения, к которому была прикреплена нажатая кнопка, клавиатура удалится
update.effective_message.edit_text(text=message_text)
```
Или использовать декоратор 
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/7bc34fa485d459a4238e72b3730bd0d01d6d5220/src/bot/conversations/common_functions/common_funcs.py#L16-L20
Который вместе с отправкой пользователю сообщения с текстом нажатой кнопки удалит клавиатуру
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/7bc34fa485d459a4238e72b3730bd0d01d6d5220/src/bot/conversations/common_functions/common_funcs.py#L25

TomatoInOil replied:
Тоже самое можно сказать про 
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/6f332ac665f44b52d02d9776735eee1d6e8cbb71/src/bot/conversations/complain/callback_funcs.py#L96-L103

TomatoInOil commented on src/bot/conversations/complain/handlers.py:
```diff
@@ -0,0 +1,79 @@
+        CallbackQueryHandler(
+            callback=common_funcs.cancel,
+            pattern=AGREE_REGEX_PATTERN.format(buttons.REPORT_NO_BUTTON),
+        ),
```
Также здесь тоже остаётся клавиатура
![image](https://github.com/Studio-Yandex-Practicum/coliving_bot/assets/93451214/b99eb141-ce4d-4ea9-8845-edd7bd438234)

Можно попробовать повесить на эту функцию декоратор, но конкретно для этого хендлера, а не для всех случаев её использования 
```Python
common_funcs.add_response_prefix()(common_funcs.cancel)
```

TomatoInOil commented on src/bot/conversations/complain/keyboards.py:
```diff
@@ -0,0 +1,76 @@
+    return screenshot_or_not_keyboard
+
+
+async def get_category_keyboard():
+    category_keyboard = InlineKeyboardMarkup.from_column(
```
Т.к. клавиатура не зависит от входных данных функции, то можно её оформить просто как константу.

TomatoInOil replied:
И здесь
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/6f332ac665f44b52d02d9776735eee1d6e8cbb71/src/bot/conversations/complain/keyboards.py#L31-L32

TomatoInOil commented on src/bot/internal_requests/entities.py:
```diff
@@ -75,6 +75,15 @@ class MatchStatuses(IntEnum):
     is_rejected = -1
 
 
+class Categories(str, Enum):
+    CATEGORY_SPAM = "Спам"
+    CATEGORY_INCORRECT_DATA = "Не совпадают личные данные"
+    CATEGORY_CHEATER = "Мошенник"
+    CATEGORY_PROHIB_ACTIV = "Запрещенная деятельность"
+    CATEGORY_BAD_LANG = "Оскорбления, мат"
+    CATEGORY_OTHER = "Другое"
```
Можно использовать для кнопок вместо этих констант, чтобы это знание не повторялось в двух местах
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/6f332ac665f44b52d02d9776735eee1d6e8cbb71/src/bot/conversations/complain/buttons.py#L9-L14

Более того по ним можно итерироваться и сократить определение клавиатуры
```Python
InlineKeyboardMarkup.from_column(
        button_column=[
            InlineKeyboardButton(text=category, callback_data=category)
            for category in Categories
        ]
    )
```

TomatoInOil commented on src/bot/conversations/complain/callback_funcs.py:
```diff
@@ -0,0 +1,150 @@
+    )
+
+
+async def handle_screenshot(update: Update, _context: ContextTypes.DEFAULT_TYPE):
+    new_screenshot = update.effective_message.photo[-1]
+    _context.user_data["complain_info"].screenshot.append(
+        Image(file_id=new_screenshot.file_id, photo_size=new_screenshot)
+    )
+    return states.RESULT
```
Т.к. мы ожидаем только 1 скриншот, то можно пропустить шаг нажатия кнопки "Сохранить фото", а сразу в этой функции продолжить логикой функции final_report для случая с изображением
https://github.com/Studio-Yandex-Practicum/coliving_bot/blob/6f332ac665f44b52d02d9776735eee1d6e8cbb71/src/bot/conversations/complain/callback_funcs.py#L118-L129
То есть диалог будет выглядеть 

> **Бот:** Есть ли у вас скриншот, на котором демонстрируется нарушение?
**Пользователь:** Нажимает кнопку "Да, я хотел бы прикрепить скриншот"
**Бот:** Отлично, скриншот поможет установить нарушение! Отправь скриншот.
**Пользователь:** Отправляет фото
**Бот:** Ваша жалоба сформирована! <...>

Тогда можно будет убрать конструкцию if в самом final_report оставив только логику, нужную для отправки сообщения без текста

TomatoInOil reviewed (changes requested):
Супер! Есть несколько небольших моментов, которые можно улучшить.

NRenat committed:
review
- Удалены лишние кнопки
- Вынесены клавиатуры для добавления скрина к жалобам и выбора категорий в константы
- Добавлено удаление кнопок после их нажатия

NRenat committed:
Добавлена обработка жалоб в backend.
- В модель UserReport добавлено поле screenshot.
- В UserReportCreateView добавлена проверка наличия пары reporter-reported_user.
- Изменена create_report, чтобы отправлять на сервер жалобы.
- Добавлена обработка повторных жалоб для пользователей.
- Добавлена обработка медии в urls.py, Debug=True.
- В админку добавлено поле preview скрина жалобы.
  - При нажатии превью увеличивается для более удобного использования админки.
  - Для открытия скриншота в новой вкладке предусмотрена кнопка "Полный размер".
- Исключено поле screenshot, так как теперь эту работу выполняет preview.
- Добавлен файл миграций.

NRenat commented on src/backend/search/views.py:
```diff
@@ -24,6 +24,26 @@ class UserReportCreateView(generics.CreateAPIView):
+            reporter=reporter, reported_user=reported_user
+        ).exists():
+            return Response(
+                {"detail": ALREADY_REPORTED}, status=status.HTTP_208_ALREADY_REPORTED
```
Не знаю насколько статус 208 подходит к этому случаю, исключая название, конечно же.
По дефолту возвращается код 400 (пара reporter-reported_user уже существует в БД), и бот падает из-за response.raise_for_status() в _post_request(). У меня не получалось вытащить причину (json ответ вместе со статусом) для обработки, поэтому возвращаю статус код 20х, который raise_for_status() пропускает.

NRenat replied:
Не знаю насколько статус 208 подходит к этому случаю, исключая название, конечно же.
По дефолту возвращается код 400 (пара reporter-reported_user уже существует в БД), и бот падает из-за response.raise_for_status() в _post_request(). У меня не получалось вытащить причину (json ответ вместе со статусом) для обработки, поэтому возвращаю статус код 20х, который raise_for_status() пропускает.

NRenat commented on src/backend/search/views.py:
```diff
@@ -24,6 +24,26 @@ class UserReportCreateView(generics.CreateAPIView):
+        reported_user = request.data.get("reported_user")
+
+        if UserReport.objects.filter(
+            reporter=reporter, reported_user=reported_user
```
Не знаю насколько статус 208 подходит к этому случаю, исключая название, конечно же.
По дефолту возвращается код 400 (пара reporter-reported_user уже существует в БД), и бот падает из-за response.raise_for_status() в _post_request(). У меня не получалось вытащить причину (json ответ вместе со статусом) для обработки, поэтому возвращаю статус код 20х, который raise_for_status() пропускает.

TomatoInOil commented on src/backend/coliving_bot/urls.py:
```diff
@@ -6,6 +7,8 @@
 )
 
+from .settings import base
+
```
https://docs.djangoproject.com/en/5.0/topics/settings/#using-settings-in-python-code

TomatoInOil commented on src/bot/conversations/complain/callback_funcs.py:
```diff
@@ -0,0 +1,169 @@
+    _context.user_data["complain_info"].category = category
+    await update.effective_message.edit_text(
+        text=ASK_COMPLAIN_TEXT,
+    )
```
Давай добавим сюда кнопку "Пропустить", если категория не равна "Другое", чтобы комментарий был необязателен в этом случае.

NRenat committed:
review
- Добавлено поле comment в админке.
- Изменено поле text, теперь оно не обязательное.
- Изменен импорт настроек через встроенные средства.

- Добавлена кнопка пропуска комментария к жалобе, если не выбрана категория "Другое".

TomatoInOil committed:
Merge commit. Resolve conflicts.

# Conflicting files
#    src/bot/conversations/coliving/callback_funcs.py
#    src/bot/internal_requests/entities.py
#    src/bot/internal_requests/service.py

TomatoInOil committed:
Squash migrations 0002 and 0003 for search app.

TomatoInOil committed:
Fix tests for new report categories in search serializers and views.
